src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
 
+static void nxt_array_elts_copy(void **dst_elts, nxt_array_t *src_array);
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
```

Comment:
> We avoid such silly comments. If something isn't clear from the function's prototype, then a bad function name has chosen. We prefer to leave comments only in non-trivial parts of the code. Developers should read code, not comments. The code is executed, the comments aren't. People make mistakes, add bugs, so no one can be 100% sure that the description of a function clearly represents how the function works. It's better if someone read the function code again before using it, rather than reading the comment.

Thank you!

Reply:
> Thank you!

Hope that I didn't accidentally offended you by the word "silly" regarding that comment.  I've actually meant an "obvious" one. Because to me it looks like:
 - This is a function to create arrays
 - Yeap, I see it, it's called array create...
 - And it returns a pointer to it!
 - No surprise here too! Its return type is `nxt_array_t *`. I can read C.

But in fact, sometimes it can also return `NULL`.

Reply:
> > Thank you!
> 
> Hope that I haven't accidentally offended you by the word "silly" regarding that comment. I've actually meant an "obvious" one. Because to me it looks like:
> 
> * This is a function to create arrays
> * Yeap, I see it, it's called array create...
> * And it returns a pointer to it!
> * No surprise here too! Its return type is `nxt_array_t *`. I can read C.
> 
> But in fact, sometimes it can also return `NULL`.

It's not my problem.

**For several reasons:**
- this is your own vision, perhaps a generally accepted principle within your team;
- I do not give such definitions to anyone's work;
- I am not an official representative and face of a public company;
- my real experience of programming in C is about 1 month and this is my first project in C in principle.

I am strictly aware of my current knowledge, skills and experience in C programming and I am aware that I can be wrong in many ways, in this regard, I am trying to get feedback from "heavyweights" in programming with a lot of experience behind them like you and yours colleagues. My principle is this: "the recognition of one's own mistakes and wrongs is the basis of personal growth."

The only thing that worries me a little is the fate of this project, at a time when such a respected and experienced specialist (without sarcasm) in C programming, accepts the current codebase as a dogma and cannot explain to me (for noob) the ultimate meaning and benefits of the chosen algorithms of memory allocation and pointer assignment by his colleagues, as well as other features of the application, based on the principles of the generally accepted technical justification laid down in all the foundations of C programming.

Reply:
The problem here is that the project you've chosen is really too complex for beginning. It has no purposes to teach someone as well.

You've probably read some books or watched some courses that showed some classical and simplified picture, that was easier to explain. And this is what you probably meant by "the principles of the generally accepted technical justification laid down in all the foundations of C programming". But they are not. These are just principles to safe beginers from mistakes in their first steps and to not blow their minds with too complexity and flexability. The reality isn't black and white: comments can be bad, changes from good to better can be bad, pointers can be used in some tricky way and point to random addresses, and "goto" operator is fine when used the right way, and etc... etc.... etc.. This all can't be explained without years of experience. The same code looks very different for a beginner and for someone with 20+ years of programming in C. Personally, I've started programming in C when I was 14yo, now I'm 35.

This is one aspect. The other aspect is that every code have some costs. I've spent time writing this message and reviewing this change, you've spent time on discussions and preparing this pull request, studying the code. Then others devs will spend time to get familiar with newer version of the code and a QA will spend time to test it. Time costs money, even if you're doing it as a hobby, then you've just pay for it by your own time and energy, that in other case could be spent on something profitable. Since everybody have to eat, have to heat their homes and pay for electricity, there are no free reviews, free pull requests, free changes, or free education. Even if it looks like, then it's just paid by someone else.

The existing code has been written by someone from the team, then has been reviewed by others, and tested by QA. And all that were paid by NGINX, Inc, part of F5.

One more aspect is that most of the code will be rewritten multiple times during project evolution, so it doesn't make sense to focus too much on polishing some existing code without clear requirements. It will be thrown away as soon as the requirements and use cases will be changed.

Another one is that there's no perfect code. The perfect code is no code at all. Any existing code is a huge compromise between performance, readability, flexibility, maintainability, and time spent on it.

And one more is that the code details don't really matter for the end user, as long as it's safe, well tested, works good and fast enough, and does what the user wants.

Taking into account all of this, would I accept and spend much time on a patch, that doesn't fix any real bug and doesn't introduce any real feature, but just makes the code look more canonical to the articles and books on C programming for beginners? No, just because the benefits aren't clear, but cost of it is far from zero, it's not well tested, and who knows, maybe it introduces some bugs, that we've failed to spot.

Look at the list in your pull request descriptions:
> removed incorrect array initialization (nxt_array_init) with a pointer to a non-existent set of elements;

Does this array initialization introduce some bug? No.
Is such pointer prohibited by C-standard, or has undefined behaviour? No, as long as it isn't dereferenced. And in this case it's not dereferenced.

So, why is it incorrect? Just because someone thinks that pointers in C should always point to only existent set of elements? They shouldn't. You can do a lot of fun stuff with pointers, but in order to do it safe, you should have some good level in C programming. Devs of this project has very good level of C programming. We don't hire with lower level.

> fixed the size of the allocated memory for storing the array structure (nxt_array_t) - waste of memory;

Does it introduce a memory leak? No.
Is it just a simple fix of dumb typo in allocation size? No.

Unused memory doesn't mean wasted memory. Allocating more memory than needed usually is used for simplifying runtime code (like in this particular case) or better aligning of structures.

> fixed initialization of a set of elements (elts), pointer, memory size;

This one looks similar to the first one. So, not clear what specifically it's about.

> fixed the formula for calculating the required amount of memory (nalloc) allocated for the created array;

Does the previous formal have bugs, e.g. giving less amount of memory than required? No. Not clear, what exactly was fixes... what is the real issue with it?

> eliminated the possibility of early deletion of an existing pointer to a set of elements (elts) at the time of memory reallocation;

Could you clarify how it affects the existing project code? Early for what? As we've discussed already since there's no concurrent access to the pointer and cannot be any concurrent access at all, since the whole API isn't thread-safe, then nohow. Current code has no such possibility and absolutely safe here.

> added descriptions of functions;

Already discussed. The best description of any function is the code of this function. Sure there are different points of view on this aspect, it depends on the team, project, programming language, and etc...

Have you seen descriptions near other similar functions? No, then no reasons to think that those descriptions will be welcome and considered as an improvement.

> added a function to get a pointer by the index of an element in an array.

Does it improve some functionality of the project? No, it doesn't. It doesn't introduce any new feature. It's not required in the existing code.

Will it be required in the future? Maybe, but should it be introduced before such feature will be implemented? No, it just some additional change to the code, that, as I've explained before, has costs.

Reply:
> You are already at least the fifth time trying to catch on to the changes I proposed, which I immediately refused at the first answer of @hongzhidao:
https://github.com/nginx/unit/pull/664#discussion_r836244623

He asked me to look at this patch. I did. If you've already refused those changes, then what we discuss?

Sorry, but I'm completely lost in what your proposal is. If we don't discuss the current change, then what change you suggest? Could you please update the pull request then.

> What is your goal by doing this?

I've tried many times to explain why that pointer is initialized to that value instead of NULL. It's initialized that way because it's a correct way of initializing nxt_array_t structure for zero number of elements, that's it.

> In many ways, you are trying to explain things that are absolutely banal and understandable to everyone (costs, time, etc.), which I told you about even earlier (https://github.com/nginx/unit/pull/664#discussion_r837444780), but for some reason you they assumed that I could not know about it.

Because this particular pull request has many "fixed" words in description and doesn't fix any bug. It just changes the code in a way that you think is better.

> Also note the fact that it is not I who can give you the technical justification for the logic I propose and explain its benefits, and you cannot do the same for your own logic. Based on your own logic, this project is difficult for you, not for me.

It's you, who proposed this pull request, not me. So, it's your obligation to give the technical justification for every single line of this change.

Reply:
> He asked me to look at this patch. I did. If you've already refused those changes, then what do we discuss?
>
> Sorry, but I'm completely lost in what your proposal is. If we don't discuss the current change, then what change do you suggest? Could you please update the pull request then.

If you understood this only from the fifth time, then you need to ask this question to yourself.

> correct way of initializing nxt_array_t

The only attempt that I consider as a real attempt by you to do this: https://github.com/nginx/unit/pull/664#discussion_r837876182
Of course there were others, but you abandoned almost all of them after my first leading question.

> Because this particular pull request has many "fixed" words in description and doesn't fix any bug. It just changes the code in a way that you think is better.
>
> It's you, who proposed this pull request, not me. So, it's your obligation to give the technical justification for every single line of this change.

I don't see anything wrong with offering a new, I think improved vision. If you do not consider it as such, then it is your responsibility as a professional with 20+ years of C programming experience to briefly and clearly explain why you think so, as @hongzhidao did it with regards to the design of the application, using only **2 sentences for this**, and **not using dozens of what seemed to me saturated with aggression comments**, to the first real attempt to at least somehow expand his vision using the technical justification for your position, without twisting previously said by me and you, without jumping from one question to another and without using other tricks in order to confuse your interlocutor.

I have stated my position regarding the pointer (`array->elts`), the abandoned allocated memory (`nxt_array_create, n> 1`), the early dereference of variables (`array->elts`) and other points in great detail and repeatedly. I think that even your colleagues will agree with this statement.

Reply:
I believe that in order to complete our debate with you, it is necessary to involve a third party. Let it be the managers of your product and company. And you and I will unconditionally accept their decision on the ideological issues raised in this discussion and will continue to rely on their vision of what is correct from the opinions voiced by both parties and whether your current position reflects the internal values ​​​​of F5 Networks.

@tippexs, @lcrilly Good afternoon! 
Sorry for the worry. Please give us some of your valuable time and give us an assessment of what is happening from your point of view. Thank you!

Reply:
> 

> This is a review thing and what @VBart did on it. This is not just for you, but also for our internal review.
> 
> > Maybe the product really lacks refactoring for the sake of refactoring or code review for the sake of code review?
> 
> First, the below picture is about the software source code quality.
> 
> ![1905745632](https://user-images.githubusercontent.com/8190423/160812238-00dbbaff-4533-4ded-8ff7-78ea25a28201.jpg)
> 
> You can see that Unit's code is pretty qualified.
> 
> Then, welcome to take a look at the insights of contributors. Our code is about 1 to 3 in terms of deletions and additions. It shows that we attach great importance to refactoring. https://github.com/nginx/unit/graphs/contributors
> 
> Sure, you are welcome and free to submit patches here. We didn't find why the patch makes sense, though we found some reasonable in your patch. This is normal for someone unfamiliar with Unit's source code.
> 
> So it's better to go back to the code itself. Unit is young, we have to choose which ones to merge and which ones to put behind. Thank you again for the issues found recently.

This is a review thing and what @VBart[](https://user-images.githubusercontent.com/8190423/160812238-00dbbaff-4533-4ded-8ff7-78ea25a28201.jpg) did on it. This is not just for you, but also for our internal review.

Maybe the product really lacks refactoring for the sake of refactoring or code review for the sake of code review?

First, the below picture is about the software source code quality.

1905745632

You can see that Unit's code is pretty qualified.

Then, welcome to take a look at the insights of contributors. Our code is about 1 to 3 in terms of deletions and additions. It shows that we attach great importance to refactoring.
https://github.com/nginx/unit/graphs/contributors

Sure, you are welcome and free to submit patches here. We didn't find why the patch makes sense, though we found some reasonable in your patch. This is normal for someone unfamiliar with Unit's source code.

So it's better to go back to the code itself.
Unit is young, we have to choose which ones to merge and which ones to put behind.
Thank you again for the issues found recently.

Reply:
This is a review thing and what @VBart[](https://user-images.githubusercontent.com/8190423/160812238-00dbbaff-4533-4ded-8ff7-78ea25a28201.jpg) did on it. This is not just for you, but also for our internal review.

Maybe the product really lacks refactoring for the sake of refactoring or code review for the sake of code review?

First, the below picture is about the software source code quality.
![1905745632](https://user-images.githubusercontent.com/8190423/160822576-b8346f4e-27f2-46bc-a1cb-82b4653af223.jpg)


You can see that Unit's code is pretty qualified.

Then, welcome to take a look at the insights of contributors. Our code is about 1 to 3 in terms of deletions and additions. It shows that we attach great importance to refactoring. https://github.com/nginx/unit/graphs/contributors

Sure, you are welcome and free to submit patches here. We didn't find why the patch makes sense, though we found some reasonable in your patch. This is normal for someone unfamiliar with Unit's source code.

So it's better to go back to the code itself. Unit is young, we have to choose which ones to merge and which ones to put behind. Thank you again for the issues found recently.

Reply:
I would like to clarify. I do not insist on the merger of this PR. The changes I propose are not vital to me.

By coincidence, the discussion that has arisen on the basis of the changes I propose can be regarded as checking the state of compliance with the publicly declared values ​​​​of F5 Networks in relation to new participants (perhaps even employees) interested in developing a separate project and the company as a whole, as well as identifying in the current team persons potentially hindering their development.

I also ask you not to correct my comments (https://github.com/nginx/unit/pull/664#discussion_r838277973) unless they violate the GitHub Community Guidelines. Each such change in my comments, I will officially record.

I asked your colleague to technically substantiate for me the logic of the current version of the code he defends, which he did not initially do, and in some places entered into an unconstructive debate.

I am potentially interested in all possible help to me (as an open source contributor) in the development of the product and I will be glad for the subsequent prosperity, otherwise I would step back from the discussion and close this PR.

Reply:
>I also ask you not to correct my comments (https://github.com/nginx/unit/pull/664#discussion_r838277973) unless they violate the GitHub Community Guidelines. Each such change in my comments, I will officially record.

It was my mistake. I commented on your message, but I have no experience in commenting Github pull request board. I knew it after I submit a comment but it covered your comment. So I tried to recover it.

Reply:
Hi @VBart and @echolimazulu,
> > removed incorrect array initialization (nxt_array_init) with a pointer to a non-existent set of elements;
> 
> Does this array initialization introduce some bug? No. Is such pointer prohibited by C-standard, or has undefined behaviour? No, as long as it isn't dereferenced. And in this case it's not dereferenced.
> 

I couldn't review the code, because the changes are huge.  Small 5 line commits with 50 lines of commit message detailing motivation and references to the ISO C Standard and other sources, and possibly testing would greatly help.

Said that, I'm concerned that in some case we might be actually invoking Undefined Behavior.   Since I haven't worked yet with that part of the code, I'm not yet familiar with it, but if I see a small commit proposing a fix to a possible UB, I'll happily review it.

I'm especially concerned because compilers are becoming more aggresive with each release, and the latest versions of the Standard disallow not only dereferencing invalid pointers, but define some cases where even storing or reading a pointer pointing to some invalid location will already invoke UB.  For example, one may not read a pointer after one past an array (which in the case of a dynamic array is not very clear if it's the requested size, or the actual allocated size), or one may not read a pointer to a deallocated (or reallocated) array after the `free(3)` or `realloc(3)` call.